File: ts/packages/luzid-grpc-client/src/client/transaction.ts
import { TransactionModificationLabeled } from '@luzid/grpc'
import {
  LabelTransactionRequest,
  LabelTransactionResponse,
  TransactionServiceClient,
  TransactionServiceDefinition,
  TransactionSubClient,
  TransactionSubDefinition,
  RpcTransactionUpdate,
  TransactionModifiedSubClient,
  TransactionModifiedSubDefinition,
} from '@luzid/grpc'
import { Channel, createClient } from '@luzid/grpc-connection'

// -----------------
// TransactionRequestClient
// -----------------
class TransactionRequestClient {
  private readonly client: TransactionServiceClient

  constructor(channel: Channel) {
    this.client = createClient(TransactionServiceDefinition, channel)
  }

  labelTransaction(
    request: LabelTransactionRequest
  ): Promise<LabelTransactionResponse> {
    return this.client.labelTransaction(request)
  }
}

// -----------------
// TransactionSubClient
// -----------------
class _TransactionSubClient {
  private readonly client: TransactionSubClient

  constructor(channel: Channel) {
    this.client = createClient(TransactionSubDefinition, channel)
  }

  subTransactions(): AsyncIterable<RpcTransactionUpdate> {
    return this.client.subTransactions({})
  }
}

// -----------------
// TransactionModifiedSubClient
// -----------------
class _TransactionModifiedSubClient {
  private readonly client: TransactionModifiedSubClient

  constructor(channel: Channel) {
    this.client = createClient(TransactionModifiedSubDefinition, channel)
  }

  subTransactionLabeled(): AsyncIterable<TransactionModificationLabeled> {
    return this.client.subTransactionLabeled({})
  }
}

// -----------------
// Consolidated TransactionClient
// -----------------
export class TransactionClient {
  private readonly request: TransactionRequestClient
  private readonly transactionSub: _TransactionSubClient
  private readonly transactionModifiedSub: _TransactionModifiedSubClient

  constructor(channel: Channel) {
    this.request = new TransactionRequestClient(channel)
    this.transactionSub = new _TransactionSubClient(channel)
    this.transactionModifiedSub = new _TransactionModifiedSubClient(channel)
  }

  labelTransaction(
    request: LabelTransactionRequest
  ): Promise<LabelTransactionResponse> {
    return this.request.labelTransaction(request)
  }

  subTransactions(): AsyncIterable<RpcTransactionUpdate> {
    return this.transactionSub.subTransactions()
  }

  subTransactionLabeled(): AsyncIterable<TransactionModificationLabeled> {
    return this.transactionModifiedSub.subTransactionLabeled()
  }
}

File: dart/luzid-grpc-client/lib/src/client/transaction.dart
import 'package:grpc/grpc_connection_interface.dart';
import 'package:luzid_grpc/luzid_grpc.dart';

// -----------------
// LabelTransactionClient
// -----------------
class LabelTransactionClient {
  final TransactionServiceClient _client;

  LabelTransactionClient(ClientChannelBase channel)
      : _client = TransactionServiceClient(channel);

  Future<LabelTransactionResponse> labelTransaction(
    LabelTransactionRequest request,
  ) {
    return _client.labelTransaction(request);
  }
}

// -----------------
// RecentTransactionUpdatesClient
// -----------------
class _RecentTransactionUpdatesClient {
  final TransactionUpdatesServiceClient _client;

  _RecentTransactionUpdatesClient(ClientChannelBase channel)
      : _client = TransactionUpdatesServiceClient(channel);

  Future<RecentTransactionUpdatesResponse> recentTransactionUpdates(
    RecentTransactionUpdatesRequest request,
  ) {
    return _client.recentTransactionUpdates(request);
  }
}

// -----------------
// TransactionSubClient
// -----------------
class _TransactionStreamSubClient {
  final TransactionSubClient _client;

  _TransactionStreamSubClient(ClientChannelBase channel)
      : _client = TransactionSubClient(channel);

  ResponseStream<RpcTransactionUpdate> subTransactions() {
    return _client.subTransactions(Empty());
  }
}

// -----------------
// TransactionModifiedSubClient
// -----------------
class _TransactionModifiedSubClient {
  final TransactionModifiedSubClient _client;

  _TransactionModifiedSubClient(ClientChannelBase channel)
      : _client = TransactionModifiedSubClient(channel);

  ResponseStream<TransactionModificationLabeled>
      subTransactionModifiedLabeled() {
    return _client.subTransactionLabeled(Empty());
  }
}

// -----------------
// Consolidated TransactionClient
// -----------------
class TransactionClient {
  final LabelTransactionClient labelTransactionClient;
  final _TransactionStreamSubClient _transactionSubClient;
  final _TransactionModifiedSubClient _transactionModifiedSubClient;
  final _RecentTransactionUpdatesClient _recentTransactionUpdatesClient;

  TransactionClient(ClientChannelBase channel)
      : labelTransactionClient = LabelTransactionClient(channel),
        _recentTransactionUpdatesClient =
            _RecentTransactionUpdatesClient(channel),
        _transactionSubClient = _TransactionStreamSubClient(channel),
        _transactionModifiedSubClient = _TransactionModifiedSubClient(channel);

  Future<LabelTransactionResponse> labelTransaction(
    LabelTransactionRequest request,
  ) {
    return labelTransactionClient.labelTransaction(request);
  }

  Future<RecentTransactionUpdatesResponse> recentTransactionUpdates(
    RecentTransactionUpdatesRequest request,
  ) {
    return _recentTransactionUpdatesClient.recentTransactionUpdates(request);
  }

  ResponseStream<RpcTransactionUpdate> subTransactions() {
    return _transactionSubClient.subTransactions();
  }

  ResponseStream<TransactionModificationLabeled>
      subTransactionModifiedLabeled() {
    return _transactionModifiedSubClient.subTransactionModifiedLabeled();
  }
}
